pkg/querier/tripperware/metadata/get_series.go

```diff
@@ -0,0 +1,289 @@
+	sp, _ := opentracing.StartSpanFromContext(ctx, "APIResponse.ToHTTPResponse")
+	defer sp.Finish()
+
+	a, ok := res.(*PrometheusGetSeriesResponseExtended)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "invalid response format")
+	}
+
+	b, err := json.Marshal(a)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "error encoding response: %v", err)
+	}
+
+	sp.LogFields(otlog.Int("bytes", len(b)))
+
+	resp := http.Response{
+		Header: http.Header{
+			"Content-Type": []string{"application/json"},
+		},
+		Body:          io.NopCloser(bytes.NewBuffer(b)),
+		StatusCode:    http.StatusOK,
+		ContentLength: int64(len(b)),
+	}
+	return &resp, nil
+}
+
+func (getSeriesCodec) MergeResponse(responses ...tripperware.Response) (tripperware.Response, error) {
+	if len(responses) == 0 {
+		return NewEmptyPrometheusGetSeriesResponse(), nil
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	promResponses := make([]*PrometheusGetSeriesResponseExtended, 0, len(responses))
+	for _, resp := range responses {
+		promResponses = append(promResponses, resp.(*PrometheusGetSeriesResponseExtended))
+	}
+
+	metricsSet := make(map[model.Fingerprint]model.LabelSet)
+	for _, promResp := range promResponses {
+		for _, metric := range promResp.Data {
+			ls := model.LabelSet{}
+			for k, v := range metric {
+				ls[model.LabelName(k)] = model.LabelValue(v)
+			}
+			fingerprint := ls.Fingerprint()
+			metricsSet[fingerprint] = ls
+		}
+	}
+
+	result := []map[string]string{}
+	for _, m := range metricsSet {
+		metric := map[string]string{}
+		for k, v := range m {
+			metric[string(k)] = string(v)
+		}
+		result = append(result, metric)
+	}
+
+	res := &PrometheusGetSeriesResponseExtended{
+		PrometheusGetSeriesResponse: PrometheusGetSeriesResponse{
+			Status: queryrange.StatusSuccess,
+		},
+		Data: result,
+	}
+	return res, nil
+}
+
+// NewEmptyPrometheusGetSeriesResponse returns an empty successful Prometheus query range response.
+func NewEmptyPrometheusGetSeriesResponse() *PrometheusGetSeriesResponseExtended {
+	return &PrometheusGetSeriesResponseExtended{
+		PrometheusGetSeriesResponse: PrometheusGetSeriesResponse{
+			Status: queryrange.StatusSuccess,
+		},
+		Data: []map[string]string{},
+	}
+}
+
+func parseTimeParam(r *http.Request, paramName string, defaultValue int64) (int64, error) {
+	val := r.FormValue(paramName)
+	if val == "" {
+		val = strconv.FormatInt(defaultValue, 10)
+	}
+	result, err := util.ParseTime(val)
+	if err != nil {
+		return 0, errors.Wrapf(err, "Invalid time value for '%s'", paramName)
+	}
+	return result, nil
+}
+
+func decorateWithParamName(err error, field string) error {
+	errTmpl := "invalid parameter %q; %v"
+	if status, ok := status.FromError(err); ok {
+		return httpgrpc.Errorf(int(status.Code()), errTmpl, field, status.Message())
+	}
+	return fmt.Errorf(errTmpl, field, err)
+}
```

Comment:
Can we reuse these methods somehow? I guess they exist for other tripperware, too